fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/LayeredDrawerWrapperImpl.java

```diff
@@ -0,0 +1,176 @@
+			iterator.add(layer);
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + afterThis + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper attachLayerBefore(Identifier beforeThis, IdentifiedLayer layer) {
+		validateUnique(layer);
+		boolean didChange = findLayer(beforeThis, (l, iterator) -> {
+			iterator.previous();
+			iterator.add(layer);
+			iterator.next();
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + beforeThis + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper removeLayer(Identifier identifier) {
+		boolean didChange = findLayer(identifier, (l, iterator) -> {
+			iterator.remove();
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + identifier + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper replaceLayer(Identifier identifier, Function<IdentifiedLayer, IdentifiedLayer> replacer) {
+		boolean didChange = findLayer(identifier, (l, iterator) -> {
+			iterator.remove();
+			iterator.add(replacer.apply((IdentifiedLayer) l));
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + identifier + " not found");
+		}
+
+		return this;
+	}
+
+	@VisibleForTesting
+	public void validateUnique(IdentifiedLayer layer) {
+		visitLayers((l, iterator) -> {
+			if (matchesIdentifier(l, layer.id())) {
+				throw new IllegalArgumentException("Layer with identifier " + layer.id() + " already exists");
+			}
+
+			return false;
+		});
+	}
+
+	@VisibleForTesting
+	public boolean findLayer(Identifier identifier, LayerVisitor visitor) {
+		AtomicBoolean found = new AtomicBoolean(false);
+
+		visitLayers((l, iterator) -> {
+			if (matchesIdentifier(l, identifier)) {
+				found.set(true);
+				return visitor.visit(l, iterator);
+			}
+
+			return false;
+		});
+
+		return found.get();
+	}
+
+	@VisibleForTesting
+	public boolean visitLayers(LayerVisitor visitor) {
+		return visitLayers(getLayers(base), visitor);
+	}
+
+	private boolean visitLayers(List<LayeredDrawer.Layer> layers, LayerVisitor visitor) {
+		AtomicBoolean modified = new AtomicBoolean(false);
+		ListIterator<LayeredDrawer.Layer> iterator = layers.listIterator();
+
+		while (iterator.hasNext()) {
+			LayeredDrawer.Layer layer = iterator.next();
+
+			if (visitor.visit(layer, iterator)) {
+				modified.set(true);
+				continue;
+			}
+
+			if (layer instanceof SubLayer subLayer) {
+				visitLayers(getLayers(subLayer.delegate()), visitor);
+			}
+		}
+
+		return modified.get();
+	}
+
+	private static boolean matchesIdentifier(LayeredDrawer.Layer layer, Identifier identifier) {
+		return layer instanceof IdentifiedLayer il && il.id().equals(identifier);
+	}
+
+	@VisibleForTesting
+	public interface LayerVisitor {
+		/**
```

Comment:
If these methods/interfaces are public, they shouldn't need `@VisibleForTesting`?

Reply:
Quite the opposite...

Reply:
Moved the test and made them package private.